File: src/main/java/com/flowcommerce/core/support/utils/DocumentService.java
package com.flowcommerce.core.support.utils;

import com.flowcommerce.core.support.error.CoreException;
import com.flowcommerce.core.support.error.ErrorType;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.UUID;

@Component
public class DocumentService {
    @Value("${file.upload.dir}")
    private String uploadDir;

    private static final long MAX_FILE_SIZE = 5L * 1024 * 1024;
    private static final List<String> ALLOWED_TYPES = List.of(
            "application/pdf",
            "image/jpeg",
            "image/png"
    );

    public FileInfo save(String directory, MultipartFile file) throws IOException {
        validateFile(file);
        String extension = getExtension(file.getOriginalFilename());
        String storedFileName = UUID.randomUUID() + extension;

        String fullPath = uploadDir + "/" + directory;
        Path directoryPath = Paths.get(fullPath);
        if (!Files.exists(directoryPath)) {
            Files.createDirectories(directoryPath);
        }

        Path filePath = directoryPath.resolve(storedFileName);
        file.transferTo(filePath);

        return new FileInfo(
                storedFileName,
                directory + "/" + storedFileName,
                file.getSize()
        );
    }

    public void delete(String filePath) throws IOException {
        Path path = Paths.get(uploadDir + "/" + filePath);
        Files.deleteIfExists(path);
    }

    private String getExtension(String filename) {
        if (filename == null || !filename.contains(".")) {
            return "";
        }
        return filename.substring(filename.lastIndexOf("."));
    }

    private void validateFile(MultipartFile file) {
        if (file == null || file.isEmpty()) {
            throw new CoreException(ErrorType.DOCUMENT_REQUIRED);
        }

        if (file.getSize() > MAX_FILE_SIZE) {
            throw new CoreException(ErrorType.DOCUMENT_SIZE_EXCEEDED);
        }

        String contentType = file.getContentType();
        if (contentType == null || !ALLOWED_TYPES.contains(contentType)) {
            throw new CoreException(ErrorType.DOCUMENT_INVALID_TYPE);
        }
    }
}

File: src/main/kotlin/com/flowcommerce/core/support/utils/DocumentService.kt
package com.flowcommerce.core.support.utils

import com.flowcommerce.core.support.error.CoreException
import com.flowcommerce.core.support.error.ErrorType
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

@Component
class DocumentService {
    @Value("\${file.upload.dir}")
    private lateinit var uploadDir: String

    companion object {
        private const val MAX_FILE_SIZE = 5L * 1024 * 1024
        private val ALLOWED_TYPES = listOf(
            "application/pdf",
            "image/jpeg",
            "image/png"
        )
    }

    @Throws(IOException::class)
    fun save(directory: String, file: MultipartFile): FileInfo {
        validateFile(file)
        val extension = getExtension(file.originalFilename)
        val storedFileName = "${UUID.randomUUID()}$extension"

        val fullPath = "$uploadDir/$directory"
        val directoryPath = Paths.get(fullPath)
        if (!Files.exists(directoryPath)) {
            Files.createDirectories(directoryPath)
        }

        val filePath = directoryPath.resolve(storedFileName)
        file.transferTo(filePath)

        return FileInfo(
            storedFileName = storedFileName,
            filePath = "$directory/$storedFileName",
            fileSize = file.size
        )
    }

    @Throws(IOException::class)
    fun delete(filePath: String) {
        val path = Paths.get("$uploadDir/$filePath")
        Files.deleteIfExists(path)
    }

    private fun getExtension(filename: String?): String {
        if (filename == null || !filename.contains(".")) {
            return ""
        }
        return filename.substring(filename.lastIndexOf("."))
    }

    private fun validateFile(file: MultipartFile) {
        if (file.isEmpty) {
            throw CoreException(ErrorType.DOCUMENT_REQUIRED)
        }

        if (file.size > MAX_FILE_SIZE) {
            throw CoreException(ErrorType.DOCUMENT_SIZE_EXCEEDED)
        }

        val contentType = file.contentType
        if (contentType == null || contentType !in ALLOWED_TYPES) {
            throw CoreException(ErrorType.DOCUMENT_INVALID_TYPE)
        }
    }
}
